cli_tools/common/utils/logging/firelog_entry.go

```diff
@@ -0,0 +1,207 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package logging
+
+import (
+	"encoding/json"
+	"fmt"
+	"runtime"
+	"time"
+)
+
+// LogRequest is a server-side pre-defined data structure
+type LogRequest struct {
+	ClientInfo    ClientInfo `json:"client_info"`
+	LogSourceName string     `json:"log_source_name"`
+	RequestTimeMs int64      `json:"request_time_ms"`
+	LogEvent      []LogEvent `json:"log_event"`
+}
+
+// ClientInfo is a server-side pre-defined data structure
+type ClientInfo struct {
+	ClientType        string            `json:"client_type"`
+	DesktopClientInfo map[string]string `json:"desktop_client_info"`
+}
+
+// LogEvent is a server-side pre-defined data structure
+type LogEvent struct {
+	EventTimeMs         int64  `json:"event_time_ms"`
+	SourceExtensionJSON string `json:"source_extension_json"`
+}
+
+// LogResponse is a server-side pre-defined data structure
+type LogResponse struct {
+	NextRequestWaitMillis int64                `json:"nextRequestWaitMillis,string"`
+	LogResponseDetails    []LogResponseDetails `json:"logResponseDetails"`
+}
+
+// LogResponseDetails is a server-side pre-defined data structure
+type LogResponseDetails struct {
+	ResponseAction ResponseAction `json:"responseAction"`
+}
+
+// ResponseAction is a server-side pre-defined data structure
+type ResponseAction string
+
+const (
+	// ResponseActionUnknown - If the client sees this, it should delete the LogRequest (not retry).
+	// It may indicate that a new response action was added, which the client
+	// doesn't yet understand.  (Deleting rather than retrying will prevent
+	// infinite loops.)  The server will do whatever it can to prevent this
+	// occurring (by not indicating an action to clients that are behind the
+	// requisite version for the action).
+	ResponseActionUnknown ResponseAction = "RESPONSE_ACTION_UNKNOWN"
+	// RetryRequestLater - The client should retry the request later, via normal scheduling.
+	RetryRequestLater ResponseAction = "RETRY_REQUEST_LATER"
+	// DeleteRequest - The client should delete the request.  This action will apply for
+	// successful requests, and non-retryable requests.
+	DeleteRequest ResponseAction = "DELETE_REQUEST"
+)
+
+func constructLogRequest(event *ComputeImageToolsLogExtension) ([]byte, error) {
+	if event == nil {
+		return nil, fmt.Errorf("won't log a nil event")
+	}
+	eventStr, err := json.Marshal(event)
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now().Unix()
+	req := LogRequest{
+		ClientInfo: ClientInfo{
+			ClientType:        "DESKTOP",
+			DesktopClientInfo: map[string]string{"os": runtime.GOOS},
+		},
+		// TODO: replace with actual log source once server side is ready: "COMPUTE_IMAGE_TOOLS"
+		LogSourceName: "FIRELOG_TEST",
+		RequestTimeMs: now,
+		LogEvent: []LogEvent{
+			{
+				EventTimeMs:         now,
+				SourceExtensionJSON: string(eventStr),
```

Comment:
Will this produce an escaped JSON string? If you're using JSON, the source_extension_json fields needs to be escaped.

Reply:
Hmm eventStr here is not escaped. However, when the whole LogReqeust is converted to JSON string, this part is definetely escaped automatically. A example:

{
   "client_info":{
      "client_type":"DESKTOP",
      "desktop_client_info":{
         "os":"linux"
      }
   },
   "log_source_name":"FIRELOG_TEST",
   "request_time_ms":1567813555,
   "log_event":[
      {
         "event_time_ms":1567813555,
         "source_extension_json":"{\"id\":\"e3efc8f0-d4e1-4cf3-93ee-63731f5249a4\",\"cloud_build_id\":\"\",\"tool_action\":\"ImageExport\",\"status\":\"Failure\",\"input_params\":{\"image_export_input_params\":{\"client_id\":\"gcloud\",\"project\":\"tzz-noogler-3\",\"obfuscated_project\":\"747a7a2d6e6f6f676c65722d33305d7251ecfb7edfdcbe8b0be7c1e4151bc797a27c86276cacee3d33c31c50bd\",\"disable_gcs_logging\":false,\"disable_cloud_logging\":false,\"disable_stdout_logging\":false,\"destination_uri\":\"gs://tzz-noogler-3-daisy-bkt/export-test-1\",\"source_image\":\"global/images/image-debian-9-20190514-2\"}},\"output_info\":{\"failure_message\":\"step \\\"image-export-export-disk\\\" run error: step \\\"wait-for-inst-image-export-export-disk\\\" run error: WaitForInstancesSignal FailureMatch found for \\\"inst-image-export-export-disk-image-export-image-export--11qx0\\\": \\\"ExportFailed: test failure  privacy_info .\\\"\",\"failure_message_without_privacy_info\":\"step %q run error: %s\"}}"
      }
   ]
}

Reply:
Looks fine.